web/profiles/custom/os2loop/modules/os2loop_toc_block/src/Helper/Helper.php

```diff
@@ -0,0 +1,131 @@
+<?php
+
+namespace Drupal\os2loop_toc_block\Helper;
+
+use Drupal\Core\Entity\Display\EntityViewDisplayInterface;
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\Core\Render\RendererInterface;
+use Drupal\node\NodeInterface;
+use Drupal\toc_api\TocBuilder;
+use Drupal\toc_api\TocManagerInterface;
+
+/**
+ * A helper.
+ */
+class Helper {
+
+  /**
+   * The toc manager interface.
+   *
+   * @var \Drupal\toc_api\TocManagerInterface
+   */
+  protected $tocManager;
+
+  /**
+   * The toc builder interface.
+   *
+   * @var \Drupal\toc_api\TocBuilderInterface
+   */
+  protected $tocBuilder;
+
+  /**
+   * The entity type manager.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected $entityTypeManager;
+
+  /**
+   * Drupal\Core\Render\RendererInterface definition.
+   *
+   * @var \Drupal\Core\Render\RendererInterface
+   */
+  protected $renderer;
+
+  /**
+   * Block constructor for table of contents.
+   *
+   * @param \Drupal\toc_api\TocManagerInterface $tocManager
+   *   Manager service for table of contents.
+   * @param \Drupal\toc_api\TocBuilder $tocBuilder
+   *   Builder service for table of contents.
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entityTypeManager
+   *   The entity manager.
+   * @param \Drupal\Core\Render\RendererInterface $renderer
+   *   Drupals renderer service.
+   */
+  public function __construct(TocManagerInterface $tocManager, TocBuilder $tocBuilder, EntityTypeManagerInterface $entityTypeManager, RendererInterface $renderer) {
+    $this->tocManager = $tocManager;
+    $this->tocBuilder = $tocBuilder;
+    $this->entityTypeManager = $entityTypeManager;
+    $this->renderer = $renderer;
+  }
+
+  /**
+   * Implements hook_ckeditor_css_alter().
+   *
+   * Adds Ids to headers related to table of contents.
+   *
+   * @param array $build
+   *   The node build.
+   * @param \Drupal\node\NodeInterface $node
+   *   The node entity.
+   * @param \Drupal\Core\Entity\Display\EntityViewDisplayInterface $display
+   *   The display of the node.
+   * @param string $view_mode
+   *   The view mode of the node.
+   *
+   * @throws \Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException
+   * @throws \Drupal\Component\Plugin\Exception\PluginNotFoundException
+   */
+  public function nodeView(array &$build, NodeInterface $node, EntityViewDisplayInterface $display, $view_mode) {
+    // Add lock to prevent the renderer from calling it self indefinitely.
+    $lock = &drupal_static(__FUNCTION__, FALSE);
+    if (!$lock) {
+      $lock = TRUE;
+      if ($view_mode == 'full') {
+        $hasToc = FALSE;
+        $paragraphs_content = $node->get('os2loop_documents_document_conte')->getValue();
```

Comment:
We should check that this field actually exists on the node or that the node is actually a document (which has this field). Otherwise this may fails or make `$paragraphs_content` be `null`.

Can we use https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Field%21EntityReferenceFieldItemList.php/function/EntityReferenceFieldItemList%3A%3AreferencedEntities/9.2.x here to get all paragraphs?